### PR TITLE
Serialize the content hash for a source text so it does not need to be recomputed

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/72703

The idea is simple here.  When sending documents over to OOP we're already computing hteir content hashes (and using that to key the document anyways).  So now, when we hydrate the document on the OOP side, we just supply the content hash directly, avoiding needing to do extra work to compute it.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Serialization/SerializableSourceText.cs [hunk 1]
```diff
@@ -28,7 +30,7 @@ internal sealed class SerializableSourceText
     /// <remarks>
     /// Exactly one of <see cref="_storage"/> or <see cref="_text"/> will be non-<see langword="null"/>.
     /// </remarks>
-    private readonly ITemporaryTextStorageWithName? _storage;
+    private readonly TemporaryTextStorage? _storage;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Serialization/SerializableSourceText.cs:33]: moved to stronger types, as several of these interfaces only have a single impl ever.

File: src/Workspaces/Core/Portable/Serialization/SerializableSourceText.cs [hunk 2]
```diff
@@ -151,31 +164,27 @@ public void Serialize(ObjectWriter writer, SolutionReplicationContext context, C
...
-            var storage = storage2.AttachTemporaryTextStorage(name, offset, size, checksumAlgorithm, encoding);
-            if (storage is ITemporaryTextStorageWithName storageWithName)
-            {
-                return new SerializableSourceText(storageWithName);
-            }
-            else
-            {
-                return new SerializableSourceText(storage.ReadText(cancellationToken));
-            }
+            var storage = storage2.AttachTemporaryTextStorage(name, offset, size, checksumAlgorithm, encoding, contentHash);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Serialization/SerializableSourceText.cs:180]: by being strongly typed, we know we know we'll always get a ITemporaryTextStorageWithName back.

File: src/Workspaces/Core/Portable/Serialization/SerializerService_Reference.cs [hunk 3]
```diff
@@ -472,7 +472,7 @@ private static ModuleMetadata ReadModuleMetadataFrom(ObjectReader reader, Serial
         }
         else
         {
-            var service2 = (ITemporaryStorageService2)_storageService;
+            var service2 = (TemporaryStorageService)_storageService;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Serialization/SerializerService_Reference.cs:475]: this was the only impl of that interface.  so any blind casts to it are safe to move to a blind cast to the class itself.

File: src/Workspaces/Remote/ServiceHub/Services/AssetSynchronization/RemoteAssetSynchronizationService.cs [hunk 4]
```diff
@@ -65,7 +65,7 @@ public ValueTask SynchronizeTextAsync(DocumentId documentId, Checksum baseTextCh
                     // Now attempt to manually apply the edit, producing the new forked text.  Store that directly in
                     // the asset cache so that future calls to retrieve it can do so quickly, without synchronizing over
                     // the entire document.
-                    var newText = new SerializableSourceText(text.WithChanges(textChanges));
+                    var newText = new SerializableSourceText(text.WithChanges(textChanges), text.GetContentHash());
                     var newChecksum = serializer.CreateChecksum(newText, cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Services/AssetSynchronization/RemoteAssetSynchronizationService.cs:69]: this is not any extra expense.  The call to serializer.CreateChecksum will force the content hash to be created anyways.

File: src/Workspaces/Core/Portable/Workspace/Host/TemporaryStorage/ITemporaryTextStorageWithName.cs [hunk 5]
```diff
@@ -1,33 +0,0 @@
...
-using System.Collections.Immutable;
-using System.Text;
-using Microsoft.CodeAnalysis.Text;
-
-namespace Microsoft.CodeAnalysis.Host;
-
-/// <summary>
-/// Represents a <see cref="ITemporaryStorageWithName"/> which is used to hold data for <see cref="SourceText"/>.
-/// </summary>
-internal interface ITemporaryTextStorageWithName : ITemporaryTextStorageInternal, ITemporaryStorageWithName
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Host/TemporaryStorage/ITemporaryTextStorageWithName.cs:14]: only one impl of this.  so this just merged into that class.

File: src/Workspaces/Remote/ServiceHub/Services/AssetSynchronization/RemoteAssetSynchronizationService.cs [hunk 6]
```diff
@@ -65,7 +65,7 @@ public ValueTask SynchronizeTextAsync(DocumentId documentId, Checksum baseTextCh
                     // Now attempt to manually apply the edit, producing the new forked text.  Store that directly in
                     // the asset cache so that future calls to retrieve it can do so quickly, without synchronizing over
                     // the entire document.
-                    var newText = new SerializableSourceText(text.WithChanges(textChanges));
+                    var newText = new SerializableSourceText(text.WithChanges(textChanges), text.GetContentHash());
```
Comment (sharwell) [inline src/Workspaces/Remote/ServiceHub/Services/AssetSynchronization/RemoteAssetSynchronizationService.cs:68]: ❓ It seems like `text.WithChanges(textChanges)` would not match `text.GetContentHash()` unless `textChanges` was empty. Are we sure this is correct?

**Review by sharwell — DISMISSED** [review]
Need a correctness verification for the question

File: src/Workspaces/Remote/ServiceHub/Services/AssetSynchronization/RemoteAssetSynchronizationService.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/ServiceHub/Services/AssetSynchronization/RemoteAssetSynchronizationService.cs]: indeed.

**Comment by CyrusNajmabadi** [comment]
@ToddGrun @sharwell @genlu ptal.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski 	For review when you get back.

**Pull request merged**